### DISCUSSION TITLE
How should F5 Execute DAX Query?

### DISCUSSION THREAD

**Discussion by mlonsk** [TabularEditor/TabularEditor3#1359] [Polls]
Based on this [feature request](https://github.com/TabularEditor/TabularEditor3/discussions/1349) we wanted to hear everyone's opinion on the best way to use F5 to execute a DAX Query.

The question is what should happen in the below scenario:

<img width="660" alt="image" src="https://github.com/user-attachments/assets/9e7ef5d0-2ae2-463a-a33f-c469e7da0884">

Current Behaviour: The whole query is run, resulting in both query results for query 1 and 2 being displayed.:
<img width="633" alt="image" src="https://github.com/user-attachments/assets/79211c9b-f878-402a-9b6a-2ddcc6805fdd">

Alternative Behaviour: Only the top query 1 is executed, showing only the results of that query.
![image](https://github.com/user-attachments/assets/502a18af-5856-4bdc-ba10-34d9aef47dbb)

**Comment by otykier** [comment]
It’s important to note that we have an alternative command today, **Shift+F5** which causes only query 1 (the selection) to be executed. This also works with the “query under the cursor” in case the current selection is empty.

**Comment by corycundy** [comment]
I don't know the effort involved, but if the default behavior doesn't get voted high enough, could there be a preference or option to have one vs. the other that can be set in preferences.

I know that while every tool will have different functionality, it certainly can be nice when there are some un-written standards that allow developers to jump from one tool to the with common functionality. Similar to how F5 is the button of choice to execute (probably from Query Analyzer in the very old days), this has become somewhat of a standard across many tools. It makes it more intuitive. I believe that is why SSMS, DAX Studio and KQL editors offer this behavior

Lastly, I don't see a big drawback in that if you want all the queries run, just ensure there is no selection. The current behavior ignores any selection. The change in required behavior if the Alternative is implemented would be to ensure there is no selection, which doesn't seem too bad.

The Alternative feels like the best of both worlds. Thanks for putting out the vote. I'm looking forward to the results!

**Reply by otykier** [reply]
Regardless of the poll outcome, we're going to add a new command, "Execute all or selection", which will run the full query (similar to F5) if no selection is made, and run the selection only (similar to Shift+F5) if a non-empty selection exists - i.e. exactly what you suggested in #1349.

The reason we're polling for this, is to decide whether we should change the default keybinding so that F5 will trigger this new "Execute all or selection" command. So even if a majority of users prefer the current functionality, you will still be able to rebind F5 to achieve exactly what you proposed :)

**Comment by otykier** [comment]
The **Execute** (F5) behavior has been modified as of 3.19.0, thanks to the outcome of this poll!

From the [release notes](https://docs.tabulareditor.com/te3/other/release-notes/3_19_0.html?tabs=downloads):

> - The **Execute** action (F5) will now execute only the selected portion of the query when text is selected. If no text is selected, the entire query will be executed.
> - **Execute Selection** (Shift+F5) will also execute the selected portion similar to the Execute action above, but when nothing is selected this action only executes the EVALUATE statement under the cursor.
> - We've introduced an **Execute full query** action (no default keybinding), in case you need a way to execute the full query without changing your current selection.

Thank you to everyone who participated in this poll and discussion!
